docs/language/informal/covariant-overrides.md

```diff
@@ -0,0 +1,742 @@
+`Widget` hierarchy was designed like that.
+
+In cases where the supertype authors did not foresee this need, it is still
+possible to tighten a parameter type by putting the `covariant` modifier on
+the *overriding* parameter declaration.
+
+In general, a tightened type for a parameter `p` in a method `m` is allowed
+when at least one of the overridden declarations of `m` has a `covariant`
+modifier on the declaration corresponding to `p`.
+
+The `covariant` modifier can also be used on mutable fields. Doing so
+corresponds to marking the parameter in the implicitly generated setter
+for that field as `covariant`:
+
+```dart
+class Widget {
+  covariant Widget child;
+}
+```
+
+This is syntactic sugar for:
+
+```dart
+class Widget {
+  Widget _child;
+  Widget get child => _child;
+  set child(covariant Widget value) { _child = value; }
+}
+```
+
+### Grammar changes
+
+We update the following grammar rules:
+
+```
+normalFormalParameter:
+  "covariant"? functionSignature |
+  fieldFormalParameter |
+  simpleFormalParameter
+  ;
+
+simpleFormalParameter:
+  metadata "covariant"? finalConstVarOrType? identifier
+  ;
+```
+
+<!--
+  Note that it is possible to mark a parameter as covariant and leave out the
+  type, e.g., `covariant var x`. This is probably useless, but benign.
+-->
+
+The `covariant` modifier can *only* be applied to a parameter declaration
+on an instance method, instance setter, or instance operator, and it can be
+applied to a mutable field declaration. Anywhere else is a static error.
+
+### Static checking
+
+The static checks are affected in several ways when adding covariant overrides
+to the language.
+
+#### Overriding
+
+With this feature, type checking of an overriding instance method or operator
+declaration depends on the `covariant` modifiers. For a given parameter
+`p` in a method or operator `m`, `p` is considered to be a **covariant
+parameter** if there is a direct or indirect supertype containing an overridden
+declaration of `m` where the parameter corresponding to `p` is marked
+`covariant`. In short, `covariant` is inherited, for each parameter. There is
+no conflict if only some overridden declarations have the `covariant` modifier
+and others do not, the parameter is covariant as soon as any one of them has
+it.
+
+We could have chosen to require that every supertype chain must make the
+parameter covariant, rather than just requiring that there exists such a
+supertype chain. This is a software engineering trade off: both choices
+can be implemented, and both choices provide a certain amount of protection
+against tightening types by accident. We have chosen the permissive variant,
+and it is always possible to make a linter require a stricter one.
+
+For a regular (non-covariant) parameter, the override rule is is unchanged:
+its type must be a supertype of the type declared for the same parameter in
+each directly overridden declaration. This includes the typical case where the
```

Comment:
The spec defines 'overrides' to mean 'overrides directly' (cf. \ref{inheritanceAndOverriding}: 'Then $m$ overrides $m^\prime$ if $m^\prime$ is not already overridden by a member of at least one of $S_1 \ldots S_{j-1}$'). This means that we can talk about 'overrides' and 'indirectly overrides'. But I thought that would be too subtle, so I made it a bit more explicit and used 'directly overrides' and 'indirectly overrides'.

Of courses, the spec never needs to talk about 'indirectly overrides', because it suffices to talk about the relationship between a declaration and the declaration that it directly overrides for all current purposes. But that won't suffice for covariant parameter type checking.